File: binary_search/1552_magnetic_force_between_two_balls/1552_magnetic_force_between_two_balls.py
from typing import List

class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()

        def check(x):
            cnt = 1
            t = position[0]
            for i in range(1, len(position)):
                if position[i] - t > x:
                    cnt += 1
                    t = position[i]
            return cnt >= m
        
        l, r = 0, position[-1]
        while l < r:
            mid = l + (r-l) // 2
            if check(mid):
                l = mid + 1
            else:
                r = mid
        return l


position = [1,2,3,4,7]
m = 3
res = Solution().maxDistance(position, m)
print(res)

File: binary_search/1552_magnetic_force_between_two_balls/1552_magnetic_force_between_two_balls.go
package main

import (
	"fmt"
	"sort"
)

func search(dist, m int, position []int) bool {
	curr, res := 0, 1
	for i := 0; i < len(position)-1; i++ {
		curr += position[i+1] - position[i]
		if curr >= dist {
			res += 1
			curr = 0
		}
	}
	return res >= m
}

func maxDistance(position []int, m int) int {
	sort.Ints(position)
	n := len(position)
	l, r := 1, position[n-1]-position[0]

	for l < r {
		mid := (l + r + 1) / 2
		if search(mid, m, position) {
			l = mid
		} else {
			r = mid - 1
		}
	}
	return l
}

func main() {
	position := []int{1, 2, 3, 4, 7}
	m := 3
	res := maxDistance(position, m)
	fmt.Println(res)
}
